---
output:
  flexdashboard::flex_dashboard:
    orientation: rows
    theme: cosmo
params: 
  topDir: 
    value: 'data/Test/'
  fusion_method:
    value: 'arriba'
  set_title:
    value: 'Patient Report'
  snv_consensus:
    value: FALSE
---
---
title: `r params$set_title`
---
```{r include=FALSE}
knitr::opts_chunk$set(comment = NA)
```

```{r read_data}
suppressPackageStartupMessages(library(knitr))
suppressPackageStartupMessages(library(plotly))
suppressPackageStartupMessages(library(kableExtra))
suppressPackageStartupMessages(library(flexdashboard))
suppressPackageStartupMessages(library(tidyverse))

# read input files
source("code/readData.R")
topDir <- params$topDir
fusion_method <- params$fusion_method
snv_consensus <- params$snv_consensus
readData(topDir = topDir, 
         fusion_method = fusion_method, 
         snv_consensus = snv_consensus)

# call driver to load other libraries/code
source("code/driver.R")
```


Key Clinical Findings (P1)
===================================== 
    
Row Page 1 {data-height=100}
-------------------------------------
   
### {data-width=300}

```{r, echo=FALSE, out.width = '100%'}
knitr::include_graphics("figures/d3b_Logo.png")
```   
 
### Patient Name
    
```{r}
cat(getPatientName());
```

### Report Date

```{r}
cat(getReportDate());
```   
 
### Report Version
    
```{r}
cat(getReportVersion());
```

Row {data-height=200}
-------------------------------------
    
### Patient & Sample Information

```{r patient_info}
x=kable(patientSampleInfo(), format="html") %>%
    kable_styling(bootstrap_options = c("bordered", "condensed", "responsive")) %>% column_spec(c(1,3,5), bold=T, border_left = "2px solid gray", border_right = "2px solid gray")
gsub("<thead>.*</thead>", "", x)
```


Row {data-height=500}
-------------------------------------
    
### Key Clinical Findings
    
```{r key_findings}
datatable(keyClinicalFindingsTable(), rownames=F);
```

Row {data-height=300}
-------------------------------------
    
### Disease Specific Information
    
```{r disease_specific_info}
dsi <- diseaseSpecificInformation()
if(nrow(dsi) > 0){
  x2=kable(diseaseSpecificInformation(), format="html") %>% 
    kable_styling(bootstrap_options = c("bordered", "condensed", "responsive")) %>% 
    column_spec(c(1), bold=T, border_left = "2px solid gray", border_right = "2px solid gray")
  gsub("<thead>.*</thead>", "", x2)
}

```

### Cancer Germline Variants

```{r}
# x2=kable(germlineOut, format="html") %>%
#   kable_styling(bootstrap_options = c("bordered", "condensed", "responsive")) %>% 
#   column_spec(c(1), bold=T, border_left = "2px solid gray", border_right = "2px solid gray")
# gsub("<thead>.*</thead>", "", x2)
```

### Genomic Summary
    
```{r genomic_summary}
gs <- genomicSummary()
if(nrow(gs) > 0){
  x2=kable(genomicSummary(), format="html") %>%
    kable_styling(bootstrap_options = c("bordered", "condensed", "responsive")) %>% 
    column_spec(c(1), bold=T, border_left = "2px solid gray", border_right = "2px solid gray")
  gsub("<thead>.*</thead>", "", x2)
}
```

   
High Confidence Alterations (P2)
=====================================     

Row Page 2  {data-height=100}
-------------------------------------
   
### {data-width=300}

```{r, echo=FALSE, out.width = '100%'}
knitr::include_graphics("figures/d3b_Logo.png")
```   
 
### Patient Name
    
```{r}
cat(getPatientName());
```

### Report Date

```{r}
cat(getReportDate());
```   
 
### Report Version
    
```{r}
cat(getReportVersion());
```

Row
-------------------------------------
### High Confidence Genomic Findings
    
```{r}
datatable(highConfidenceFindingsTable(), rownames=F);
```

Row 
-------------------------------------
    
### Transcriptome-Proteome Alterations
    
```{r plot_genes}
if(exists('expData')){
  ggplotly(plotGenes()); 
}
```

### Pathway Alterations

    
```{r plot_pathway}
#Bar chart of genes
if(exists('expData')){
  ggplotly(plotPathway());
}
```

Tumor Burden/Signatures (P3)
=====================================     

Row Page 3  {data-height=100}
-------------------------------------
   
### {data-width=300}

```{r, echo=FALSE, out.width = '100%'}
knitr::include_graphics("figures/d3b_Logo.png")
```   
 
### Patient Name
    
```{r}
cat(getPatientName());
```

### Report Date

```{r}
cat(getReportDate());
```   
 
### Report Version
    
```{r}
cat(getReportVersion());
```

Row
-------------------------------------
### Tumor Mutational Signatures
    
```{r}
if(exists('mutData')){
  ggplotly(tumorSignaturePlot())
}
```

### Tumor Mutational Burden
    
```{r fig.width = 11, fig.height=7}
ggplotly(tmbProfile())
```

Tumor Purity (P4)
=====================================     

Row Page 4  {data-height=100}
-------------------------------------
   
### {data-width=300}

```{r, echo=FALSE, out.width = '100%'}
knitr::include_graphics("figures/d3b_Logo.png")
```   
 
### Patient Name
    
```{r}
cat(getPatientName());
```

### Report Date

```{r}
cat(getReportDate());
```   
 
### Report Version
    
```{r}
cat(getReportVersion());
```

Row 
-------------------------------------
    
### Immune Profile
    
```{r immune_profile, fig.width = 28, fig.height=14}
if(exists('expData')){
  p <- ImmuneProfile()
  ggplotly(p)
}
```


Genomically Similar Samples (P5)
=====================================     

Row Page 5  {data-height=100}
-------------------------------------
   
   
### {data-width=300}

```{r, echo=FALSE, out.width = '100%'}
knitr::include_graphics("figures/d3b_Logo.png")
```   
 
### Patient Name
    
```{r}
cat(getPatientName());
```

### Report Date

```{r}
cat(getReportDate());
```   
 
### Report Version
    
```{r}
cat(getReportVersion());
```

Row
-------------------------------------
### Nearest-Neighbor Analysis
    
```{r tsne_plot}
if(exists('expData')){
  ggplotly(getTSNEPlot())
}
```


Row 
-------------------------------------
    
### Survival
    
```{r km_plot, fig.width = 6}
if(exists('expData')){
  getKMPlot();
}
```

### Treatment/Outcomes - Genomically Similar Patients
    
```{r similar_patients}
datatable(getSimilarPatients(), rownames=F, escape=F);
```


All Findings (P6)
=====================================     

Row Page 6  {data-height=100}
-------------------------------------
   
### {data-width=300}

```{r, echo=FALSE, out.width = '100%'}
knitr::include_graphics("figures/d3b_Logo.png")
```   
 
### Patient Name
    
```{r}
cat(getPatientName());
```

### Report Date

```{r}
cat(getReportDate());
```   
 
### Report Version
    
```{r}
cat(getReportVersion());
```

Row {data-height=900}
-------------------------------------
### All Findings
    
```{r all_findings}
datatable(allFindingsTable(), rownames=F, options = list(pageLength = 10));
```

Genomic Landscape (P7)
=====================================     

Row Page 7  {data-height=100}
-------------------------------------
   
### {data-width=300}

```{r, echo=FALSE, out.width = '100%'}
knitr::include_graphics("figures/d3b_Logo.png")
```   
 
### Patient Name
    
```{r}
cat(getPatientName());
```

### Report Date

```{r}
cat(getReportDate());
```   
 
### Report Version
    
```{r}
cat(getReportVersion());
```

Row
-------------------------------------
### Circos
    
```{r circos, fig.width = 7, fig.height = 6}
if(exists('fusData')){
 knitr::include_graphics(paste0(topDir, "/tmpRCircos.png")) 
}
```

### Network View
    
```{r plot_network, fig.width = 8, fig.height = 6}
if(exists('expData')){
  plotNetwork()
}
```

Row 
-------------------------------------
    
### Genome CNA View
    
```{r plot_cnv, fig.width = 20}
# Plot Genome Wide CNV
if(exists("cnvData")){
  plotCNV()
}
```