---
output:
  flexdashboard::flex_dashboard:
    orientation: rows
    theme: cosmo
params: 
  topDir: 
    value: 'data/Test/'
  fusion_method:
    value: 'arriba'
  set_title:
    value: 'Patient Report'
  snv_pattern:
    value: NULL
  tmb:
    value: NULL
editor_options: 
  chunk_output_type: console
---
---
title: `r params$set_title`
---

```{r include=FALSE, echo=FALSE}
knitr::opts_chunk$set(comment = NA)
```

```{r load_libraries}
source('code/load_libraries.R')
```

```{r read_data}
# read input files
source("code/readData.R")
topDir <- params$topDir
fusion_method <- params$fusion_method
snv_pattern <- params$snv_pattern
tmb <- params$tmb
readData(topDir = topDir, 
         fusion_method = fusion_method, 
         snv_pattern = snv_pattern)
```

```{r call_driver}
# call driver to load other libraries/code
source("code/driver.R")
```

Key Clinical Findings (P1)
===
Row {data-height=200}
---   
### {data-width=200}

```{r, echo=FALSE, out.width = '100%'}
knitr::include_graphics("figures/d3b_Logo.png")
```
    
### Patient & Sample Information

```{r patient_info}
x=kable(patientSampleInfo(), format="html") %>%
    kable_styling(bootstrap_options = c("bordered", "condensed", "responsive")) %>% column_spec(c(1,3), bold=T, border_left = "2px solid gray", border_right = "2px solid gray")
gsub("<thead>.*</thead>", "", x)
```

Row {data-height=500}
---
    
### Key Clinical Findings
    
```{r key_findings}
datatable(keyClinicalFindingsTable(), rownames=F, escape = F);
```

Row {data-height=300}
---
    
### Disease Specific Information
    
```{r disease_specific_info}
dsi <- diseaseSpecificInformation(snv_pattern)
if(nrow(dsi) > 0){
  x2=kable(dsi, format="html") %>% 
    kable_styling(bootstrap_options = c("bordered", "condensed", "responsive")) %>% 
    column_spec(c(1), bold=T, border_left = "2px solid gray", border_right = "2px solid gray")
  gsub("<thead>.*</thead>", "", x2)
}

```

### Cancer Germline Variants

```{r germline_vars}
if(exists('mutData.germ')){
  germlineOut <- filterGermline()
  if(nrow(germlineOut) >= 1){
    kable(germlineOut, format="html") %>%
      kable_styling(bootstrap_options = c("bordered", "condensed", "responsive")) %>%
      column_spec(c(1), bold=T, border_left = "2px solid gray", border_right = "2px solid gray")
  }
}
```

### Genomic Summary
    
```{r genomic_summary}
gs <- genomicSummary(snv_pattern)
if(nrow(gs) > 0){
  x2=kable(gs, format="html") %>%
    kable_styling(bootstrap_options = c("bordered", "condensed", "responsive")) %>% 
    column_spec(c(1), bold=T, border_left = "2px solid gray", border_right = "2px solid gray")
  gsub("<thead>.*</thead>", "", x2)
}
```

Transcriptomic Alterations (P2)
===
Row {data-height=700}
---

### Transcriptome Alterations
    
```{r plot_genes}
if(exists('expData')){
  ggplotly(plotGenes()); 
}
```

### Pathway Alterations

```{r plot_pathway}
#Bar chart of genes
if(exists('expData')){
  ggplotly(plotPathway());
}
```

Tumor Burden/Signatures (P3)
===
Row {data-height=700}
---

### Tumor Mutational Signatures
    
```{r tm_signature}
if(exists('mutData')){
  ggplotly(tumorSignaturePlot())
}
```

### Tumor Mutational Burden
    
```{r tm_burden, fig.width = 11, fig.height=7}
if(!is.null(tmb)){
  ggplotly(tmbProfile(TMB = tmb))  
}
```

Immune Profile: Pediatric (P4) {data-navmenu="Tumor Purity"}
===
Row {data-height=700}
---
    
### Immune Profile (with PBTA)
    
```{r immune_profile_pbta, fig.width = 28, fig.height=14}
if(exists('expData')){
  p <- ImmuneProfile(fullmat = pbta.mat.all, 
                     fname = paste0(topDir,'/ImmuneScores/rawScores_pediatric.txt'))
  ggplotly(p)
}
```

Immune Profile: Adult (P4) {data-navmenu="Tumor Purity"}
===
Row {data-height=700}
---
    
### Immune Profile (with TCGA GBM)
    
```{r immune_profile_tcga, fig.width = 28, fig.height=14}
if(exists('expData')){
  p <- ImmuneProfile(fullmat = tcga.gbm.mat, 
                     fname = paste0(topDir,'/ImmuneScores/rawScores_adult.txt'))
  ggplotly(p)
}
```

TIS Profile (P4) {data-navmenu="Tumor Purity"}
===
Row {data-height=700}
---
    
### TIS Profile (TCGA + PBTA)
    
```{r tis_profile, fig.width = 28, fig.height=14}
if(exists('expData')){
  p <- tisProfile(fname = paste0(topDir,'/ImmuneScores/tisScores.txt'), 
                  score = 'avg')
  ggplotly(p)
}
```

Genomically Similar Samples (PBTA) (P5) {data-navmenu="Pediatric Tumors Analysis"}
===
Row
---

### Nearest-Neighbor Analysis
    
```{r dimred_plot_pbta}
if(exists('expData')){
  getDimRedPlot(dat = pbta.embedding, 
                clindata = pbta.clinData, 
                study = "PBTA", 
                patient = sampleInfo$subjectID, 
                title =  "UMAP Correlation Clustering")
}
```

Row 
---
    
### Survival
    
```{r km_plot_pbta, fig.width = 6}
if(exists('expData')){
  getKMPlot(allCor = pbta.allCor, survData = pbta.survData)
}
```

### Treatment/Outcomes - Genomically Similar Patients
    
```{r similar_patients_pbta}
datatable(getSimilarPatients(allCor = pbta.allCor, clinData = pbta.clinData), rownames=F, escape=F);
```

Tumor Purity (P7) {data-navmenu="Pediatric Tumors Analysis"}
===
Row {data-height=700}
---
    
### Immune Profile (Genomically Similar Patients)
    
```{r immune_profile_gs, fig.width = 28, fig.height=14}
if(exists('expData')){
  p <- ImmuneProfile(fullmat = pbta.topCor, fname = paste0(topDir,'/ImmuneScores/topCor_rawScores.txt'))
  ggplotly(p)
}
```

Pathway Enrichment (P8) {data-navmenu="Pediatric Tumors Analysis"}
===
Row {data-height=500}
---

### Pathway Enrichment (Genomically Similar Patients)
    
```{r tabulate_pathways_pediatric}
gsPath <- tabulate_pathways(allCor = pbta.allCor, prefix = "pediatric", comparison = "pediatric")
datatable(gsPath[[1]], rownames=F, options = list(pageLength = 10));
```

Row {data-height=500}
---

### CNV Mapping 

```{r cnv_mapping_pediatric}
datatable(gsPath[[2]], rownames=F, options = list(pageLength = 10));
```

ssGSEA (P9) {data-navmenu="Pediatric Tumors Analysis"}
===
Row {data-height=700}
---
    
### ssGSEA (PBTA HGAT samples)
    
```{r ssgsea_hgg, fig.width = 28, fig.height=14}
if(exists('expData')){
  p <- ssGSEA(topCor = pbta.mat.hgg, fname = paste0(topDir,'/GSVA/ssgsea_rawScores.txt'))
  ggplotly(p)
}
```

Mutational Analysis (P13) {data-navmenu="Pediatric Tumors Analysis"}
===
Row {data-height=500}
---

### Recurrently Altered Genes (Genomically Similar Patients)
    
```{r rec_alt_genes_pediatric}
total.alt.table <<- recurrent.alterations(topCor = pbta.topCor)
datatable(total.alt.table[[1]], rownames=F, options = list(pageLength = 10));
```

Row {data-height=500}
---

### Shared Genes 

```{r shared_genes_pediatric}
datatable(total.alt.table[[2]], rownames=F, options = list(pageLength = 10));
```

Genomically Similar Samples (TCGA) (P6) {data-navmenu="Adult Tumors Analysis"}
===
Row
---

### Nearest-Neighbor Analysis
    
```{r dimred_plot_tcga}
if(exists('expData')){
  getDimRedPlot(dat = tcga.gbm.embedding, 
                clindata = tcga.gbm.clinData, 
                study = "TCGA", 
                patient = sampleInfo$subjectID, 
                title =  "UMAP Correlation Clustering")
}
```

Row 
---
    
### Survival
    
```{r km_plot_tcga, fig.width = 6}
if(exists('expData')){
  getKMPlot(allCor = tcga.gbm.allCor, survData = tcga.gbm.survData)
}
```

### Treatment/Outcomes - Genomically Similar Patients
    
```{r similar_patients_tcga}
datatable(getSimilarPatients(allCor = tcga.gbm.allCor, clinData = tcga.gbm.clinData), rownames=F, escape=F);
```

Pathway Enrichment (P8) {data-navmenu="Adult Tumors Analysis"}
===
Row {data-height=500}
---

### Pathway Enrichment (Genomically Similar Patients)
    
```{r tabulate_pathways_adult}
gsPath <- tabulate_pathways(allCor = tcga.gbm.allCor, prefix = "adult", comparison = "adult")
datatable(gsPath[[1]], rownames=F, options = list(pageLength = 10));
```

All Findings (P10)
===
Row {data-height=700}
---

### All Findings
    
```{r all_findings}
datatable(allFindingsTable(snv_pattern), rownames=F, escape = F, options = list(pageLength = 10));
```

Genomic Landscape (P11)
===
Row
---

### Circos
    
```{r circos, fig.width = 7, fig.height = 6}
if(exists('fusData')){
 knitr::include_graphics(paste0(topDir, "/tmpRCircos.png")) 
}
```

### Network View
    
```{r plot_network, fig.width = 8, fig.height = 6}
if(exists('expData') | exists('mutData') | exists('fusData')){
  plotNetwork()
}
```

Row 
---
    
### Genome CNA View
    
```{r plot_cnv, fig.width = 20}
# Plot Genome Wide CNV
if(exists("cnvData")){
  plotCNV()
}
```

PHGG: Expression and CNV Heatmap  (P12) {data-navmenu="Heatmaps"}
===
Row {data-height=700}
---

### Expression-CNV Heatmap (PHGG)
    
```{r heatmap_phgg, echo=FALSE}
fname <- paste0(topDir,'/complexHeatmap_phgg.png')
knitr::include_graphics(fname)
```

CGS: Expression and CNV Heatmap (P12) {data-navmenu="Heatmaps"}
===
Row
---

### Expression-CNV Heatmap (CGS)

```{r heatmap_cgs, echo=FALSE}
fname <- paste0(topDir,'/complexHeatmap_cgs.png')
knitr::include_graphics(fname) 
```

Oncogrid (HGG-DMG)  (P14)
===
Row {data-height=700}
---
  
### Oncogrid (HGG-DMG)
  
```{r oncogrid_hgg_dmg, echo=FALSE}
fname <- file.path(topDir,'complexHeatmap_oncogrid.png')
knitr::include_graphics(fname)
```